### ISSUE TITLE
Traefik forward TCP connection to closed TCP port

### ISSUE THREAD

**Issue by ezh** [containous/traefik#5598]
Traefik forward connections even if there is no open TCP port.
In general there are a lot of cases when you may export more then one TCP port from single container or when container is up but specific port is still (or already) unavailable.

### What did you expect to see?
Some kind of TCP health check or `traefik.tcp.middlewares.test-retry.retry.attempts` or anything other that will provide more reliable service.

### What did you see instead?
A lot of errors...

### Output of `traefik version`: (_What version of Traefik are you using?_)
```
traefik version
Version:      2.0.1
Codename:     montdor
Go version:   go1.13.1
Built:        2019-09-26T16:18:03Z
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

docker-compose.yml with 1 load balancer and 3 nodes.
Configuration of nodes is absolutely the same except name.
```
version: '3.0'                                                                                                                           services:
  lb:
    hostname: lb
    image: traefik:latest
    command:
      - --api.insecure=true
      - --providers.docker
      - --entryPoints.http.address=:80
      - --entryPoints.kubeapi.address=:6443
      - --log.level=DEBUG
    ports:
      - "6443:6443"
      - "8080:8080"
      - "80:80"
    labels:
      - "traefik.enable=false"
    restart: always
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
  node1:
    build:
      context: ./config
    hostname: node1
    image: kube
    labels:
      - traefik.tcp.routers.kube-apiserver.entrypoints=kubeapi
      - traefik.tcp.routers.kube-apiserver.rule=HostSNI(`*`)
      - traefik.tcp.routers.kube-apiserver.service=kube-apiserver
      - traefik.tcp.routers.kube-apiserver.tls.passthrough=true
      - traefik.tcp.services.kube-apiserver.loadbalancer.server.port=6443
    privileged: true
    tty: true
    restart: always
    volumes:
      - /sys/fs/cgroup:/sys/fs/cgroup:rw
      - ./config:/config
  node2:
     ...
```

No other configurations.

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)

```
lb_1     | time="2019-10-08T12:25:36Z" level=debug msg="Handling connection from 172.25.0.4:54836"
lb_1     | time="2019-10-08T12:25:36Z" level=error msg="Error while connection to backend: dial tcp 172.25.0.3:6443: connect: connection refused"
lb_1     | time="2019-10-08T12:25:36Z" level=debug msg="Handling connection from 172.25.0.4:54846"
lb_1     | time="2019-10-08T12:25:36Z" level=error msg="Error while connection to backend: dial tcp 172.25.0.5:6443: connect: connection refused"
lb_1     | time="2019-10-08T12:25:36Z" level=debug msg="Handling connection from 172.25.0.4:54850"
lb_1     | time="2019-10-08T12:25:36Z" level=debug msg="Error while terminating connection: close tcp 172.25.0.2:6443->172.25.0.4:54850: shutdown:
transport endpoint is not connected"
```

**Comment by Damu112** [comment]
+1. In the current version, we cannot use Traefik for TCP load balancing because unhealthy nodes stay in the distribution, thus breaking the application.

**Comment by dbsanfte** [comment]
This is a critical missing feature.

**Comment by povils** [comment]
Any progress regarding this issue?

**Comment by mwolf1989** [comment]
This would be a big enhancement. A similar behavior for tcp routes to http routes would be great.

**Comment by zwtop** [comment]
Hello, is there any plan in which version to add the feature (TCP loadbalancer health check)?

**Comment by ricardo-rod** [comment]
It will be great for the load balancer to have TCP and UDP to have health checks, :+1:

**Comment by StephanSalas** [comment]
I really think this should be implemented... I'm currently working on https://github.com/traefik/traefik/issues/8541 and I noticed the same. I think something as is described in AWS' documentation would make this more feature-complete: 

> In this case, a health check simply attempts to open a TCP connection to the instance on the specified port. Failure to connect within the configured timeout is considered unhealthy.

Here's the issue: This is going to require us to update the following:

The dynamic objects, adding a healthcheck interface that solely works with TCP Conns, and a UI Dashboard change to support the same. This is a LOT of work, which could also introduce a whole set of other bugs/issues. But I think the team needs to consider this as every other major cloud loadbalancer supports this, and Traefik is falling behind.

**Comment by gfonseca-tc** [comment]
Would also really appreciate having a TCP healthcheck, since traefik does not support HTTP POST healthchecks neither. It really reduces the possibilities of what you can do with it.

**Comment by dark4igi** [comment]
any updates?
it's must have feature for loadbalancer solution/software.

**Comment by nergdron** [comment]
just ran into this as well. I've got a service that exposes an http endpoint with a health check, and then it also has extra tcp ports I wanna pass through, but there seems to be no way to connect the http health check to the tcp router / service.

**Comment by t0mtaylor** [comment]
Would be great to have this in 2023! TCP Healthchecks are pretty straight forward, shouldn't take too long to implement 👍

**Comment by FabianKnapp** [comment]
Do it! :)

**Comment by dark4igi** [comment]
> Do it! :)

Where?
In what version it's available?
Who to test it?

**Comment by rtribotte** [comment]
Hello,

We would love some help from the community on this, if any community member would like to contribute to this, let us know, and we will work with you to make sure you have all the information needed before starting, and ensure that we are aligned and can move quickly with the review and merge process.

**Comment by ddtmachado** [comment]
@rtribotte I also started looking at this some time ago, when I needed to handle TCP HC. I'm moving forward slowly whenever time allows so if you would like we could sync up on what's missing and if the direction is right. It differs from @eos175 as I'm not using its library and cutting out the last ACK, plus I'm creating a separate health check config for TCP as I believe its different enough to deserve that. I would also like to support a simple send X / receive Y expectation as an option. WIP [here](https://github.com/ddtmachado/traefik/tree/tcp-hc)

**Comment by rtribotte** [comment]
Hi @ddtmachado,

Good to hear from you, hope you are doing well! And good news! I'll try to take a look in the next few weeks and give you feedback, we could then sync up.

**Comment by rtribotte** [comment]
Hey @ddtmachado,

We took a look at the changes with @kevinpollet, and long story short, you can totally open a PR right now, even in draft.
For us it is the most easy way to iterate on the topic.

At a glance, we think the server TLS config should be used with the healthcheck dialer to allow TLS.
We noticed the payload and expected option, and this could have been postpone for later improvement, but why not.

The address option doesn't seem to be necessary, as we'll use the service one. Otherwise, and it seems that you started this way, the TCP healthcheck mechanism should be overall more or less a copy of the HTTP one.

**Comment by eos175** [comment]
Hello!
The method it uses is intrusive it fully opens a TCP connection, which means the kernel will hand over the connection to the program. By not sending the final ACK, the kernel never passes the connection to the program, so it remains unaware that someone attempted to connect. 

:/

**Comment by ddtmachado** [comment]
@rtribotte @kevinpollet awesome! I'll open the PR as draft right now. I just added tests with the payload / expected option and will work on the TLS and other missing parts, hopefully this week. Have a good one! 🍻

**Comment by ddtmachado** [comment]
> Hello! The method it uses is intrusive it fully opens a TCP connection, which means the kernel will hand over the connection to the program. By not sending the final ACK, the kernel never passes the connection to the program, so it remains unaware that someone attempted to connect.
> 
> :/

@eos175 I do not understood the issue in opening a full TCP connection. To me at least, it seems the right thing to do when validating my TCP server is working as expected and accepting new connections, or even answering with some specific bytes.

Can you describe on a scenario were stopping at the kernel would make more sense?

**Comment by eos175** [comment]
@ddtmachado  If I want to respond to the request, I would use HTTP or configure the program to send a request directly to my Uptime. TCP operates on layer 4 and is designed for specific protocols, so I believe it shouldn’t be intrusive, as each protocol has its own characteristics. This is exactly what HAProxy does.

Additionally, passing the connection to the program would fill the logs with connection-disconnection records, which might be unnecessary. In fact, I initially implemented it this way: [link](https://github.com/traefik/traefik/commit/0b91ca73a80497fbae92b2f63018470a2c84f9e2#diff-2c565121f665a1bc64eef7d8aeb4c3c406473167bb5fb93426755d18c98fa204R169)
